### ISSUE TITLE
node-typescript-compiler: Failure during tsc invocation: Spawn: got event "err" (on Windows only)

### ISSUE THREAD

**Issue by photonstorm** [Offirmo/offirmo-monorepo#5]
I've been using node-typescript-compiler v2.2.1 quite happily on Mac OS, but on Windows 10 it always errors, using the exact same tsconfig, input files, etc. Calling tsc directly from the Win10 cmd works fine (passing it the same config file and input options), it only errors from node.

Here's a verbose output:

```
node-typescript-compiler v2.2.1:
node-typescript-compiler: spawning: "D:\wamp\www\phaser-fx\node_modules\typescript\bin\tsc --moduleResolution node --target esnext --module es2015 --lib DOM,DOM.Iterable,ES5,ES6,ESNext,ScriptHost --emitDeclarationOnly --noEmitOnError --sourceMap --declaration --declarationMap --removeComments --esModuleInterop --downlevelIteration --emitDecoratorMetadata --experimentalDecorators --allowSyntheticDefaultImports --alwaysStrict --noImplicitAny --noImplicitThis --strictBindCallApply --strictFunctionTypes --suppressImplicitAnyIndexErrors --outDir Swap Post FX/dist/ --declarationDir Swap Post FX/dist/ Swap Post FX/src/SwapPostFX.ts"

[node-typescript-compiler] Failure during tsc invocation: Spawn: got event "err"
Error: spawn D:\wamp\www\phaser-fx\node_modules\typescript\bin\tsc ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:269:19)
    at onErrorNT (internal/child_process.js:465:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'spawn D:\\wamp\\www\\phaser-fx\\node_modules\\typescript\\bin\\tsc',
  path: 'D:\\wamp\\www\\phaser-fx\\node_modules\\typescript\\bin\\tsc',
  spawnargs: [
    '--moduleResolution',
    'node',
    '--target',
    'esnext',
    '--module',
    'es2015',
    '--lib',
    'DOM,DOM.Iterable,ES5,ES6,ESNext,ScriptHost',
    '--emitDeclarationOnly',
    '--noEmitOnError',
    '--sourceMap',
    '--declaration',
    '--declarationMap',
    '--removeComments',
    '--esModuleInterop',
    '--downlevelIteration',
    '--emitDecoratorMetadata',
    '--experimentalDecorators',
    '--allowSyntheticDefaultImports',
    '--alwaysStrict',
    '--noImplicitAny',
    '--noImplicitThis',
    '--strictBindCallApply',
    '--strictFunctionTypes',
    '--suppressImplicitAnyIndexErrors',
    '--outDir',
    'Swap Post FX/dist/',
    '--declarationDir',
    'Swap Post FX/dist/',
    'Swap Post FX/src/SwapPostFX.ts'
  ],
  stdout: '',
  stderr: '',
  reason: 'Spawn: got event "err"'
}
Error: spawn D:\wamp\www\phaser-fx\node_modules\typescript\bin\tsc ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:269:19)
    at onErrorNT (internal/child_process.js:465:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'spawn D:\\wamp\\www\\phaser-fx\\node_modules\\typescript\\bin\\tsc',
  path: 'D:\\wamp\\www\\phaser-fx\\node_modules\\typescript\\bin\\tsc',
  spawnargs: [
    '--moduleResolution',
    'node',
    '--target',
    'esnext',
    '--module',
    'es2015',
    '--lib',
    'DOM,DOM.Iterable,ES5,ES6,ESNext,ScriptHost',
    '--emitDeclarationOnly',
    '--noEmitOnError',
    '--sourceMap',
    '--declaration',
    '--declarationMap',
    '--removeComments',
    '--esModuleInterop',
    '--downlevelIteration',
    '--emitDecoratorMetadata',
    '--experimentalDecorators',
    '--allowSyntheticDefaultImports',
    '--alwaysStrict',
    '--noImplicitAny',
    '--noImplicitThis',
    '--strictBindCallApply',
    '--strictFunctionTypes',
    '--suppressImplicitAnyIndexErrors',
    '--outDir',
    'Swap Post FX/dist/',
    '--declarationDir',
    'Swap Post FX/dist/',
    'Swap Post FX/src/SwapPostFX.ts'
  ],
  stdout: '',
  stderr: '',
  reason: 'Spawn: got event "err"'
}
[node-typescript-compiler] Failure during tsc invocation: Spawn: got event "close" with error code "-4058" & signal "null"
Error: Spawn: got event "close" with error code "-4058" & signal "null"
    at on_failure (D:\wamp\www\phaser-fx\node_modules\node-typescript-compiler\src\compile.js:53:17)
    at ChildProcess.<anonymous> (D:\wamp\www\phaser-fx\node_modules\node-typescript-compiler\src\compile.js:108:8)
    at ChildProcess.emit (events.js:315:20)
    at maybeClose (internal/child_process.js:1048:16)
    at Socket.<anonymous> (internal/child_process.js:439:11)
    at Socket.emit (events.js:315:20)
    at Pipe.<anonymous> (net.js:673:12) {
  stdout: '',
  stderr: '',
  reason: 'Spawn: got event "close" with error code "-4058" & signal "null"'
}
```

It's being called from a function like this:

```js
        await tsc.compile(
            {
                ...tsconfig.compilerOptions,
                'outDir': dist,
                'declarationDir': dist
            },
            [
                src + shortName + '.ts'
            ],
            {
                verbose: true
            }
        );
```

This is wrapped in a try-catch, but essentially this is it.

Any ideas? I've seen a couple of closed issues with the same error, but no resolution in them, so figured it might be safer to start a new issue with the verbose output. Could it be as simple as not handling the Windows backslashes?

**Comment by Offirmo** [comment]
Hi @photonstorm ,

thanks for the report and for using my tool! Yes, you made the right call opening a new issue with `verbose: true`, this is very helpful!

I think the error message is pretty clear: node-typescript-compiler is trying to invoke `D:\\wamp\\www\\phaser-fx\\node_modules\\typescript\\bin\\tsc` but there is no file there (`ENOENT`)

Unfortunately, I don't have a Window machine. However, we can debug together.

Could you check the content of `D:\\wamp\\www\\phaser-fx\\node_modules\\typescript\\bin`? Maybe Windows is slightly different, for example I suspect the script file may be named `tsc.bat`

**Comment by Offirmo** [comment]
@photonstorm I just published a minor update v2.3.0 which may improve Windows compatibility and some logs. No other code change. Please try it!

**Comment by photonstorm** [comment]
I can confirm that `D:\wamp\www\phaser-fx\node_modules\typescript\bin\tsc` does indeed exist.

Here is the script running again under 2.3.0:

```
[node-typescript-compiler] v2.3.0:
[node-typescript-compiler] ✔ found a typescript compiler at this location: "D:\wamp\www\phaser-fx\node_modules\typescript\bin\tsc"
[node-typescript-compiler] ► now spawning the compilation command: "D:\wamp\www\phaser-fx\node_modules\typescript\bin\tsc --moduleResolution node --target esnext --module es2015 --lib DOM,DOM.Iterable,ES5,ES6,ESNext,ScriptHost --emitDeclarationOnly --noEmitOnError --sourceMap --declaration --declarationMap --removeComments --esModuleInterop --downlevelIteration --emitDecoratorMetadata --experimentalDecorators --allowSyntheticDefaultImports --alwaysStrict --noImplicitAny --noImplicitThis --strictBindCallApply --strictFunctionTypes --suppressImplicitAnyIndexErrors --outDir Swap Post FX/dist/ --declarationDir Swap Post FX/dist/ Swap Post FX/src/SwapPostFX.ts"...

[node-typescript-compiler] ✖ Failure during tsc invocation: Spawn: got event "err"
Error: spawn D:\wamp\www\phaser-fx\node_modules\typescript\bin\tsc ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:269:19)
    at onErrorNT (internal/child_process.js:465:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'spawn D:\\wamp\\www\\phaser-fx\\node_modules\\typescript\\bin\\tsc',
  path: 'D:\\wamp\\www\\phaser-fx\\node_modules\\typescript\\bin\\tsc',
  spawnargs: [
    '--moduleResolution',
    'node',
    '--target',
    'esnext',
    '--module',
    'es2015',
    '--lib',
    'DOM,DOM.Iterable,ES5,ES6,ESNext,ScriptHost',
    '--emitDeclarationOnly',
    '--noEmitOnError',
    '--sourceMap',
    '--declaration',
    '--declarationMap',
    '--removeComments',
    '--esModuleInterop',
    '--downlevelIteration',
    '--emitDecoratorMetadata',
    '--experimentalDecorators',
    '--allowSyntheticDefaultImports',
    '--alwaysStrict',
    '--noImplicitAny',
    '--noImplicitThis',
    '--strictBindCallApply',
    '--strictFunctionTypes',
    '--suppressImplicitAnyIndexErrors',
    '--outDir',
    'Swap Post FX/dist/',
    '--declarationDir',
    'Swap Post FX/dist/',
    'Swap Post FX/src/SwapPostFX.ts'
  ],
  stdout: '',
  stderr: '',
  reason: 'Spawn: got event "err"'
}
Error: spawn D:\wamp\www\phaser-fx\node_modules\typescript\bin\tsc ENOENT
    at Process.ChildProcess._handle.onexit (internal/child_process.js:269:19)
    at onErrorNT (internal/child_process.js:465:16)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'spawn D:\\wamp\\www\\phaser-fx\\node_modules\\typescript\\bin\\tsc',
  path: 'D:\\wamp\\www\\phaser-fx\\node_modules\\typescript\\bin\\tsc',
  spawnargs: [
    '--moduleResolution',
    'node',
    '--target',
    'esnext',
    '--module',
    'es2015',
    '--lib',
    'DOM,DOM.Iterable,ES5,ES6,ESNext,ScriptHost',
    '--emitDeclarationOnly',
    '--noEmitOnError',
    '--sourceMap',
    '--declaration',
    '--declarationMap',
    '--removeComments',
    '--esModuleInterop',
    '--downlevelIteration',
    '--emitDecoratorMetadata',
    '--experimentalDecorators',
    '--allowSyntheticDefaultImports',
    '--alwaysStrict',
    '--noImplicitAny',
    '--noImplicitThis',
    '--strictBindCallApply',
    '--strictFunctionTypes',
    '--suppressImplicitAnyIndexErrors',
    '--outDir',
    'Swap Post FX/dist/',
    '--declarationDir',
    'Swap Post FX/dist/',
    'Swap Post FX/src/SwapPostFX.ts'
  ],
  stdout: '',
  stderr: '',
  reason: 'Spawn: got event "err"'
}
[node-typescript-compiler] ✖ Failure during tsc invocation: Spawn: got event "close" with error code "-4058" & signal "null"
Error: Spawn: got event "close" with error code "-4058" & signal "null"
    at on_failure (D:\wamp\www\phaser-fx\node_modules\node-typescript-compiler\src\compile.js:54:17)
    at ChildProcess.<anonymous> (D:\wamp\www\phaser-fx\node_modules\node-typescript-compiler\src\compile.js:110:8)
    at ChildProcess.emit (events.js:315:20)
    at maybeClose (internal/child_process.js:1048:16)
    at Socket.<anonymous> (internal/child_process.js:439:11)
    at Socket.emit (events.js:315:20)
    at Pipe.<anonymous> (net.js:673:12) {
  stdout: '',
  stderr: '',
  reason: 'Spawn: got event "close" with error code "-4058" & signal "null"'
}
```

**Comment by Offirmo** [comment]
Googling it, it seems it's a common issue https://www.google.com/search?q=windows+node+js+spawn+enoent

which is not easily troubleshooted without access to a Windows machine...

I don't think I'll be able to fix it anytime soon.

**Comment by photonstorm** [comment]
Leave it with me, I'll create a basic exec test next week and try it out.

**Comment by photonstorm** [comment]
Just to say that I've found a solution for this problem of it not working on Windows.

If you use the package [cross-spawn](https://www.npmjs.com/package/cross-spawn) it works perfectly!

Just add the package and then replace:

```js
const { spawn } = require('child_process')
```

with

```
const spawn = require('cross-spawn')
```

in `compile.js` and that's all it needs :) Works flawlessly for me after that on Mac OS and Windows!

**Comment by Offirmo** [comment]
Thanks @photonstorm ! Very good suggestion, I'll try to make the change as soon as I can (no promise)

**Comment by Offirmo** [comment]
Hi @photonstorm thanks for the suggestion and sorry for the delay. I just released a v3.0.0 using cross-spawn as you suggested.
